src/lambda/wizards/samInitWizard.ts

```diff
@@ -1,552 +1,257 @@
+        if (!value) {
+            return localize('AWS.samcli.initWizard.name.error.empty', 'Application name cannot be empty')
         }
 
-        ;[this.runtime, this.packageType, this.dependencyManager] = result
-        if (this.dependencyManager === undefined) {
-            return WIZARD_RETRY
+        if (value.includes(path.sep)) {
+            return localize(
+                'AWS.samcli.initWizard.name.error.pathSep',
+                'The path separator ({0}) is not allowed in application names',
+                path.sep
+            )
         }
 
-        return wizardContinue(this.TEMPLATE)
+        return undefined
     }
 
-    private readonly TEMPLATE: WizardStep = async () => {
-        this.template = await this.context.promptUserForTemplate(this.runtime!, this.packageType!)
-
-        if (this.template === repromptUserForTemplate) {
-            return WIZARD_RETRY
-        }
-        if (this.template === eventBridgeStarterAppTemplate) {
-            return wizardContinue(this.REGION)
-        }
-
-        return this.template ? wizardContinue(this.LOCATION) : WIZARD_GOBACK
-    }
+    return createInputBox({
+        value: defaultValue,
+        title: localize('AWS.samcli.initWizard.name.prompt', 'Enter a name for your new application'),
+        buttons: makeButtons(samInitDocUrl),
+        validateInput: validateName,
+    })
+}
 
-    private readonly REGION: WizardStep = async () => {
-        this.region = await this.context.promptUserForRegion()
+export class CreateNewSamAppWizard extends Wizard<CreateNewSamAppWizardForm> {
+    public constructor(
+        context: {
+            samCliVersion: string
+            regionProvider: RegionProvider
+            awsContext: AwsContext
+            credentials?: AWS.Credentials
+        },
+        initState?: CreateNewSamAppWizardForm
+    ) {
+        super()
 
-        return this.region ? wizardContinue(this.REGISTRY) : WIZARD_GOBACK
-    }
+        this.form.runtimeAndPackage.bindPrompter(() => createRuntimePrompter(context.samCliVersion))
 
-    private readonly REGISTRY: WizardStep = async () => {
-        this.registryName = await this.context.promptUserForRegistry(this.region!)
+        this.form.dependencyManager.bindPrompter(state => createDependencyPrompter(state.runtimeAndPackage!.runtime!), {
+            showWhen: state =>
+                state.runtimeAndPackage?.runtime !== undefined &&
+                getDependencyManager(state.runtimeAndPackage.runtime).length > 1,
+            setDefault: state =>
+                state.runtimeAndPackage?.runtime !== undefined
+                    ? getDependencyManager(state.runtimeAndPackage.runtime)[0]
+                    : undefined,
+        })
 
-        return this.registryName ? wizardContinue(this.SCHEMA) : WIZARD_GOBACK
-    }
+        this.form.template.bindPrompter(state =>
+            createTemplatePrompter(
+                state.runtimeAndPackage!.runtime!,
```

Comment:
Note that callbacks manipulating the wizard state currently use a lot of a 'no-null' assertion operators since the `showWhen` clauses are effectively a black box. I have some updates to the framework that resolves this and makes it much cleaner, but it's not quite done yet.

Reply:
Control flow is being tested so nothing bad could really happen. Basically the types aren't smart enough to infer that the property is bound in this version, but it's very obvious from the code that they should be.

Reply:
I have the code for updating the types which makes the control flow truly declarative, but I think it's best to just merge this first as the benefit is relatively small.